The database and redis are deployed separately The container failed to start while starting harbor

harbor version: 1.10.1
Apr 1 00:23:16 localhost core[12065]: goroutine 1 [running]:
Apr 1 00:23:16 localhost core[12065]: github.com/astaxie/beego.initBeforeHTTPRun()
Apr 1 00:23:16 localhost core[12065]: #11/harbor/src/vendor/github.com/astaxie/beego/beego.go:104 +0x20d
Apr 1 00:23:16 localhost core[12065]: github.com/astaxie/beego.Run(0x0, 0x0, 0x0)
Apr 1 00:23:16 localhost core[12065]: #11/harbor/src/vendor/github.com/astaxie/beego/beego.go:58 +0x26
Apr 1 00:23:16 localhost core[12065]: main.main()
Apr 1 00:23:16 localhost core[12065]: #11/harbor/src/core/main.go:292 +0xeb0
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [ERROR] [/common/config/manager.go:118]: loadSystemConfigFromEnv failed, config item, key: clair_db_port, err: strconv.Atoi: parsing "": invalid syntax
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/native/adapter.go:42]: the factory for adapter docker-registry registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/harbor/adapter.go:40]: the factory for adapter harbor registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/dockerhub/adapter.go:25]: Factory for adapter docker-hub registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/huawei/huawei_adapter.go:27]: the factory of Huawei adapter was registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/googlegcr/adapter.go:29]: the factory for adapter google-gcr registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/awsecr/adapter.go:47]: the factory for adapter aws-ecr registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/azurecr/adapter.go:15]: Factory for adapter azure-acr registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/aliacr/adapter.go:26]: the factory for adapter ali-acr registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/jfrog/adapter.go:30]: the factory of jfrog artifactory adapter was registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/quayio/adapter.go:34]: the factory of Quay.io adapter was registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/helmhub/adapter.go:30]: the factory for adapter helm-hub registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/gitlab/adapter.go:19]: the factory for adapter gitlab registered
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/controllers/base.go:289]: Config path: /etc/core/app.conf
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:177]: initializing configurations...
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/config/config.go:100]: key path: /etc/core/key
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [ERROR] [/common/config/manager.go:118]: loadSystemConfigFromEnv failed, config item, key: clair_db_port, err: strconv.Atoi: parsing "": invalid syntax
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/config/config.go:73]: init secret store
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/config/config.go:76]: init project manager based on deploy mode
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/config/config.go:145]: initializing the project manager based on local database...
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:181]: configurations initialization completed
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/common/dao/base.go:84]: Registering database: type-PostgreSQL host-172.31.0.23 port-20010 databse-registry sslmode-"disable"
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/common/dao/base.go:89]: Register database completed
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/common/dao/pgsql.go:119]: Upgrading schema for pgsql ...
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/common/dao/pgsql.go:122]: No change in schema, skip.
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:83]: User id: 1 already has its encrypted password.
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:245]: initializing notification...
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/pkg/notification/notification.go:50]: notification initialization completed
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:269]: Because SYNC_REGISTRY set false , no need to sync registry
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:272]: Init proxy
Apr 1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:291]: Version: v1.10.1, Git commit: f3e1171
Apr 1 00:24:17 localhost core[12065]: panic: ERR Client sent AUTH, but no password is set
Apr 1 00:24:17 localhost core[12065]:
Apr 1 00:24:17 localhost core[12065]: goroutine 1 [running]:
Apr 1 00:24:17 localhost core[12065]: github.com/astaxie/beego.initBeforeHTTPRun()
Apr 1 00:24:17 localhost core[12065]: #11/harbor/src/vendor/github.com/astaxie/beego/beego.go:104 +0x20d
Apr 1 00:24:17 localhost core[12065]: github.com/astaxie/beego.Run(0x0, 0x0, 0x0)
Apr 1 00:24:17 localhost core[12065]: #11/harbor/src/vendor/github.com/astaxie/beego/beego.go:58 +0x26
Apr 1 00:24:17 localhost core[12065]: main.main()
Apr 1 00:24:17 localhost core[12065]: #11/harbor/src/core/main.go:292 +0xeb0

Apr  1 00:23:16 localhost core[12065]: goroutine 1 [running]:
Apr  1 00:23:16 localhost core[12065]: github.com/astaxie/beego.initBeforeHTTPRun()
Apr  1 00:23:16 localhost core[12065]: #011/harbor/src/vendor/github.com/astaxie/beego/beego.go:104 +0x20d
Apr  1 00:23:16 localhost core[12065]: github.com/astaxie/beego.Run(0x0, 0x0, 0x0)
Apr  1 00:23:16 localhost core[12065]: #011/harbor/src/vendor/github.com/astaxie/beego/beego.go:58 +0x26
Apr  1 00:23:16 localhost core[12065]: main.main()
Apr  1 00:23:16 localhost core[12065]: #011/harbor/src/core/main.go:292 +0xeb0
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [ERROR] [/common/config/manager.go:118]: loadSystemConfigFromEnv failed, config item, key: clair_db_port,  err: strconv.Atoi: parsing "": invalid syntax
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/native/adapter.go:42]: the factory for adapter docker-registry registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/harbor/adapter.go:40]: the factory for adapter harbor registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/dockerhub/adapter.go:25]: Factory for adapter docker-hub registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/huawei/huawei_adapter.go:27]: the factory of Huawei adapter was registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/googlegcr/adapter.go:29]: the factory for adapter google-gcr registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/awsecr/adapter.go:47]: the factory for adapter aws-ecr registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/azurecr/adapter.go:15]: Factory for adapter azure-acr registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/aliacr/adapter.go:26]: the factory for adapter ali-acr registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/jfrog/adapter.go:30]: the factory of jfrog artifactory adapter was registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/quayio/adapter.go:34]: the factory of Quay.io adapter was registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/helmhub/adapter.go:30]: the factory for adapter helm-hub registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/replication/adapter/gitlab/adapter.go:19]: the factory for adapter gitlab registered
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/controllers/base.go:289]: Config path: /etc/core/app.conf
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:177]: initializing configurations...
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/config/config.go:100]: key path: /etc/core/key
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [ERROR] [/common/config/manager.go:118]: loadSystemConfigFromEnv failed, config item, key: clair_db_port,  err: strconv.Atoi: parsing "": invalid syntax
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/config/config.go:73]: init secret store
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/config/config.go:76]: init project manager based on deploy mode
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/config/config.go:145]: initializing the project manager based on local database...
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:181]: configurations initialization completed
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/common/dao/base.go:84]: Registering database: type-PostgreSQL host-172.31.0.23 port-20010 databse-registry sslmode-"disable"
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/common/dao/base.go:89]: Register database completed
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/common/dao/pgsql.go:119]: Upgrading schema for pgsql ...
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/common/dao/pgsql.go:122]: No change in schema, skip.
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:83]: User id: 1 already has its encrypted password.
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:245]: initializing notification...
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/pkg/notification/notification.go:50]: notification initialization completed
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:269]: Because SYNC_REGISTRY set false , no need to sync registry 
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:272]: Init proxy
Apr  1 00:24:17 localhost core[12065]: 2020-03-31T16:24:17Z [INFO] [/core/main.go:291]: Version: v1.10.1, Git commit: f3e11715
Apr  1 00:24:17 localhost core[12065]: panic: ERR Client sent AUTH, but no password is set
Apr  1 00:24:17 localhost core[12065]: 
Apr  1 00:24:17 localhost core[12065]: goroutine 1 [running]:
Apr  1 00:24:17 localhost core[12065]: github.com/astaxie/beego.initBeforeHTTPRun()
Apr  1 00:24:17 localhost core[12065]: #011/harbor/src/vendor/github.com/astaxie/beego/beego.go:104 +0x20d
Apr  1 00:24:17 localhost core[12065]: github.com/astaxie/beego.Run(0x0, 0x0, 0x0)
Apr  1 00:24:17 localhost core[12065]: #011/harbor/src/vendor/github.com/astaxie/beego/beego.go:58 +0x26
Apr  1 00:24:17 localhost core[12065]: main.main()
Apr  1 00:24:17 localhost core[12065]: #011/harbor/src/core/main.go:292 +0xeb0

--- a/make/photon/prepare/utils/configs.py
+++ b/make/photon/prepare/utils/configs.py
@@ -360,8 +360,17 @@ def get_redis_configs(external_redis=None, with_clair=True):
     >>> get_redis_configs()['redis_url_clair']
     'redis://redis:6379/4'
 
+    >>> get_redis_configs({'host': 'localhost', 'password': ''})['redis_password']
+    ''
+    >>> get_redis_configs({'host': 'localhost', 'password': None})['redis_password']
+    ''
+    >>> get_redis_configs({'host': 'localhost', 'password': None})['redis_url_reg']
+    'redis://localhost:6379/1'
+
     >>> get_redis_configs({'host': 'localhost', 'password': 'pass'})['external_redis']
     True
+    >>> get_redis_configs({'host': 'localhost', 'password': 'pass'})['redis_password']
+    'pass'
     >>> get_redis_configs({'host': 'localhost', 'password': 'pass'})['redis_url_reg']
     'redis://anonymous:pass@localhost:6379/1'
     >>> get_redis_configs({'host': 'localhost', 'password': 'pass'})['redis_url_js']
@@ -372,6 +381,7 @@ def get_redis_configs(external_redis=None, with_clair=True):
     >>> 'redis_url_clair' not in get_redis_configs(with_clair=False)
     True
     """
+    external_redis = external_redis or {}
 
     configs = dict(external_redis=bool(external_redis))
 
@@ -387,7 +397,7 @@ def get_redis_configs(external_redis=None, with_clair=True):
     }
 
     # overwriting existing keys by external_redis
-    redis.update(external_redis or {})
+    redis.update({key: value for (key, value) in external_redis.items() if value})
 
     configs['redis_host'] = redis['host']
     configs['redis_port'] = redis['port']

